// @iobroker/json-config/build/JsonConfigComponent/ConfigCheckDocker.js
import React from 'react';
import { FormControlLabel, Checkbox, FormHelperText, FormControl, CircularProgress } from '@mui/material';
import { I18n } from '@iobroker/adapter-react-v5';
import ConfigGeneric from './ConfigGeneric';
// Send to admin message to check if the docker available and if available, show checkbox, else show warning if specified in the schema
class ConfigCheckDocker extends ConfigGeneric {
    async componentDidMount() {
        super.componentDidMount();
        const id = await this.props.oContext.socket.getCurrentInstance();
        this.setState({ requesting: true }, async () => {
            const result = await this.props.oContext.socket.sendTo(id, 'checkDocker', null);
            this.setState({
                requesting: false,
                version: result?.version || '',
                errorDocker: result
                    ? !result.daemonRunning
                        ? I18n.t('ra_Docker is not installed or not running')
                        : ''
                    : I18n.t('ra_No response from admin'),
            });
        });
    }
    renderItem(error, disabled) {
        const value = ConfigGeneric.getValue(this.props.data, this.props.attr);
        if (this.state.requesting) {
            return React.createElement(CircularProgress, null);
        }
        if (this.state.errorDocker && !value) {
            return (React.createElement(FormHelperText, { style: { color: 'orange' } },
                I18n.t('ra_Docker is not available'),
                ":\u00A0",
                I18n.t(`ra_${this.state.errorDocker}`).replace(/^ra_/, '')));
        }
        return (React.createElement(FormControl, { style: { width: '100%' }, variant: "standard" },
            React.createElement(FormControlLabel, { onClick: e => {
                    e.preventDefault();
                    e.stopPropagation();
                    if (!disabled) {
                        const mayByPromise = this.onChange(this.props.attr, !value);
                        if (mayByPromise instanceof Promise) {
                            void mayByPromise.catch(e => console.error(`Cannot set value: ${e}`));
                        }
                    }
                }, control: React.createElement(Checkbox, { checked: !!value, onChange: e => {
                        void this.onChange(this.props.attr, e.target.checked);
                    }, disabled: disabled || (!!this.state.errorDocker && !value) }), label: this.getText(this.props.schema.label) }),
            !this.props.schema.hideVersion ? (React.createElement(FormHelperText, null, this.state.errorDocker ? this.state.errorDocker : this.state.version)) : null,
            this.props.schema.help ? (React.createElement(FormHelperText, null, this.renderHelp(this.props.schema.help, this.props.schema.helpLink, this.props.schema.noTranslation))) : null));
    }
}
export default ConfigCheckDocker;
