// @iobroker/json-config/build/JsonConfigComponent/ConfigTable.js
import React, { createRef } from 'react';
import Dropzone from 'react-dropzone';
import { Accordion, AccordionDetails, AccordionSummary, Button, Card, Dialog, DialogActions, DialogContent, DialogTitle, Grid2, IconButton, InputAdornment, Paper, Table, TableBody, TableCell, TableContainer, TableHead, TableRow, TableSortLabel, TextField, Toolbar, Tooltip, Typography, FormHelperText, } from '@mui/material';
import { Add as AddIcon, Delete as DeleteIcon, Close as CloseIcon, ArrowUpward as UpIcon, ArrowDownward as DownIcon, FilterAlt as IconFilterOn, FilterAltOff as IconFilterOff, ContentCopy as CopyContentIcon, Download as ExportIcon, Warning as ErrorIcon, UploadFile as ImportIcon, Close as IconClose, ExpandMore as ExpandMoreIcon, } from '@mui/icons-material';
import { I18n } from '@iobroker/adapter-react-v5';
import ConfigGeneric from './ConfigGeneric';
import ConfigPanel from './ConfigPanel';
const MAX_SIZE = 1024 * 1024; // 1MB
const styles = {
    fullWidth: {
        width: '100%',
    },
    root: {
        width: '100%',
    },
    paper: {
        width: '100%',
        marginBottom: 16,
        backgroundColor: 'rgba(255, 255, 255, 0.1)',
    },
    headerText: {
        width: '100%',
    },
    table: {
        minWidth: 750,
    },
    visuallyHidden: {
        border: 0,
        clip: 'rect(0 0 0 0)',
        height: 1,
        margin: -1,
        overflow: 'hidden',
        padding: 0,
        position: 'absolute',
        top: 20,
        width: 1,
    },
    label: {
        display: 'flex',
        justifyContent: 'space-between',
    },
    // highlight: (theme: IobTheme): React.CSSProperties => (theme.palette.mode === 'light'
    //     ? {
    //         color: theme.palette.secondary.main,
    //         // backgroundColor: lighten(theme.palette.secondary.light, 0.85),
    //     }
    //     : {
    //         color: theme.palette.text.primary,
    //         backgroundColor: theme.palette.secondary.dark,
    //     }),
    title: {
        flex: '1 1 100%',
    },
    rootTool: {
        paddingLeft: 16,
        paddingRight: 8,
    },
    silver: {
        opacity: 0.2,
    },
    flex: {
        display: 'flex',
        alignItems: 'baseline',
    },
    filteredOut: {
        padding: 10,
        display: 'flex',
        textAlign: 'center',
    },
    buttonEmpty: {
        width: 34,
        display: 'inline-block',
    },
    buttonCell: {
        whiteSpace: 'nowrap',
    },
    dropZone: {
        width: '100%',
        height: 100,
        position: 'relative',
    },
    dropZoneEmpty: {},
    uploadDiv: {
        position: 'relative',
        width: '100%',
        minHeight: 300,
        opacity: 0.9,
        marginTop: 30,
        cursor: 'pointer',
        outline: 'none',
    },
    uploadDivDragging: {
        opacity: 1,
        background: 'rgba(128,255,128,0.1)',
    },
    image: {
        objectFit: 'contain',
        margin: 'auto',
        display: 'flex',
        width: '100%',
        height: '100%',
    },
    uploadCenterDiv: {
        margin: 5,
        border: '3px dashed grey',
        borderRadius: 5,
        width: 'calc(100% - 10px)',
        height: 'calc(100% - 10px)',
        minHeight: 300,
        position: 'relative',
        display: 'flex',
    },
    uploadCenterIcon: {
        paddingTop: 10,
        width: 48,
        height: 48,
    },
    uploadCenterText: {
        fontSize: 16,
    },
    uploadCenterTextAndIcon: {
        textAlign: 'center',
        position: 'absolute',
        top: 0,
        bottom: 0,
        left: 0,
        right: 0,
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
    },
    buttonRemoveWrapper: {
        position: 'absolute',
        zIndex: 222,
        right: 0,
    },
    error: {
        border: '2px solid red',
        boxSizing: 'border-box',
    },
    tooltip: {
        pointerEvents: 'none',
    },
};
function objectToArray(object, nameOfFirstAttr, nameOfSecondAttr) {
    nameOfFirstAttr = nameOfFirstAttr || 'key';
    const array = [];
    Object.keys(object).forEach(key => {
        const item = {};
        item[nameOfFirstAttr] = key;
        if (nameOfSecondAttr) {
            item[nameOfSecondAttr] = object[key];
            array.push(item);
        }
        else {
            array.push(Object.assign(item, object[key]));
        }
    });
    return array;
}
function arrayToObject(array, nameOfFirstAttr, nameOfSecondAttr) {
    nameOfFirstAttr = nameOfFirstAttr || 'key';
    const object = {};
    array.forEach((row) => {
        let key = row[nameOfFirstAttr];
        if (key === null || key === undefined) {
            key = '';
        }
        delete row[nameOfFirstAttr];
        if (nameOfSecondAttr) {
            object[key] = row[nameOfSecondAttr];
        }
        else {
            object[key] = row;
        }
    });
    return object;
}
function encrypt(secret, value) {
    let result = '';
    for (let i = 0; i < value.length; i++) {
        result += String.fromCharCode(secret[i % secret.length].charCodeAt(0) ^ value.charCodeAt(i));
    }
    return result;
}
function decrypt(secret, value) {
    let result = '';
    for (let i = 0; i < value.length; i++) {
        result += String.fromCharCode(secret[i % secret.length].charCodeAt(0) ^ value.charCodeAt(i));
    }
    return result;
}
class ConfigTable extends ConfigGeneric {
    filterRefs;
    typingTimer = null;
    resizeTimeout = null;
    secret = 'Zgfr56gFe87jJOM';
    refDiv;
    constructor(props) {
        super(props);
        this.filterRefs = {};
        this.props.schema.items = this.props.schema.items || [];
        this.props.schema.items.forEach((el) => {
            if (el.filter) {
                this.filterRefs[el.attr] = createRef();
            }
        });
        this.refDiv = React.createRef();
    }
    /**
     * React lifecycle hook, called once as component is mounted
     */
    async componentDidMount() {
        super.componentDidMount();
        const _value = ConfigGeneric.getValue(this.props.data, this.props.attr) || [];
        let value;
        // if the list is given as an object
        if (this.props.schema.objKeyName) {
            value = objectToArray(_value, this.props.schema.objKeyName, this.props.schema.objValueName);
        }
        else {
            value = _value;
        }
        if (!Array.isArray(value)) {
            value = [];
        }
        if (this.props.schema.encryptedAttributes) {
            const systemConfig = await this.props.oContext.socket.getCompactSystemConfig();
            this.secret = systemConfig?.native.secret || this.secret;
            _value.forEach((el) => {
                this.props.schema.encryptedAttributes.forEach((attr) => {
                    if (el[attr]) {
                        el[attr] = decrypt(this.secret, el[attr]);
                    }
                });
            });
        }
        this.setState({
            value,
            visibleValue: null,
            orderBy: /* this.props.schema.items.length ? this.props.schema.items[0].attr : */ '',
            order: 'asc',
            iteration: 0,
            filterOn: [],
            width: 0,
            tableErrors: {},
        }, () => this.validateUniqueProps());
    }
    componentWillUnmount() {
        if (this.typingTimer) {
            clearTimeout(this.typingTimer);
            this.typingTimer = null;
        }
        if (this.resizeTimeout) {
            clearTimeout(this.resizeTimeout);
            this.resizeTimeout = null;
        }
        super.componentWillUnmount();
    }
    itemTable(attrItem, data, idx) {
        const { schema } = this.props;
        const schemaForAttribute = schema.items?.find((el) => el.attr === attrItem);
        if (!schemaForAttribute) {
            return null;
        }
        const schemaItem = {
            items: {
                [attrItem]: schemaForAttribute,
            },
        };
        return (React.createElement(ConfigPanel, { oContext: this.props.oContext, alive: this.props.alive, arrayIndex: idx, changed: this.props.changed, common: this.props.common, expertMode: this.props.expertMode, custom: true, data: data, globalData: this.props.data, index: idx + this.state.iteration, onChange: (attr, valueChange) => {
                const newObj = JSON.parse(JSON.stringify(this.state.value));
                newObj[idx][attr] = valueChange;
                this.setState({ value: newObj }, () => {
                    this.validateUniqueProps();
                    this.onChangeWrapper(newObj, true);
                });
            }, onError: this.onTableRowError(idx), originalData: this.props.originalData, schema: schemaItem, table: true, themeName: this.props.themeName }));
    }
    /**
     * Validate that columns configured in `uniqueColumns` have unique values
     */
    validateUniqueProps() {
        if (!this.props.schema.uniqueColumns) {
            return;
        }
        let firstErrorColumn = null;
        let firstErrorValue = null;
        // Validate all columns and collect errors
        for (const uniqueCol of this.props.schema.uniqueColumns) {
            const allVals = [];
            const found = this.state.value.find(entry => {
                const val = entry[uniqueCol];
                if (allVals.includes(val)) {
                    // Store the first error we encounter
                    if (!firstErrorColumn) {
                        firstErrorColumn = uniqueCol;
                        firstErrorValue = val;
                    }
                    this.onError(uniqueCol, 'is not unique');
                    return true;
                }
                allVals.push(val);
                return false;
            });
            // Clear error for this column if no duplicates found
            if (!found) {
                this.onError(uniqueCol, null);
            }
        }
        // Set error message based on the first error found (or clear if no errors)
        if (firstErrorColumn) {
            this.setState({
                errorMessage: I18n.t('Non-allowed duplicate entry "%s" in column "%s"', firstErrorValue, firstErrorColumn),
            });
        }
        else {
            this.setState({ errorMessage: '' });
        }
    }
    /**
     * Handle errors from table row components
     */
    onTableRowError = (rowIndex) => (attr, error) => {
        const newTableErrors = { ...this.state.tableErrors };
        if (!newTableErrors[rowIndex]) {
            newTableErrors[rowIndex] = {};
        }
        if (!error) {
            delete newTableErrors[rowIndex][attr];
            // Clean up empty row error objects
            if (Object.keys(newTableErrors[rowIndex]).length === 0) {
                delete newTableErrors[rowIndex];
            }
        }
        else {
            newTableErrors[rowIndex][attr] = error;
        }
        this.setState({ tableErrors: newTableErrors });
        // Forward error to parent component
        this.props.onError(attr, error);
    };
    static descendingComparator(a, b, orderBy) {
        if (b[orderBy] < a[orderBy]) {
            return -1;
        }
        if (b[orderBy] > a[orderBy]) {
            return 1;
        }
        return 0;
    }
    static getComparator(order, orderBy) {
        return order === 'desc'
            ? (a, b) => ConfigTable.descendingComparator(a, b, orderBy)
            : (a, b) => -ConfigTable.descendingComparator(a, b, orderBy);
    }
    static getFilterValue(el) {
        return el?.current?.children[0]?.children[0]?.value;
    }
    static setFilterValue(el, filterValue) {
        return (el.current.children[0].children[0].value = filterValue);
    }
    handleRequestSort = (property, orderCheck = false) => {
        const { order, orderBy } = this.state;
        //if (orderBy || 'asc') {
        const isAsc = orderBy === property && order === 'asc';
        const newOrder = orderCheck ? order : isAsc ? 'desc' : 'asc';
        const newValue = this.stableSort(newOrder, property);
        // update only if value is really changed
        if (JSON.stringify(newValue) !== JSON.stringify(this.state.value)) {
            this.setState({ value: newValue, order: newOrder, orderBy: property, iteration: this.state.iteration + 10000 }, () => this.applyFilter(false, newValue));
        }
        //}
    };
    stableSort = (order, orderBy) => {
        const { value } = this.state;
        const comparator = ConfigTable.getComparator(order, orderBy);
        const stabilizedThis = value.map((el, index) => ({ el, index }));
        stabilizedThis.sort((a, b) => {
            const order_ = comparator(a.el, b.el);
            if (order_ !== 0) {
                return order_;
            }
            return a.index - b.index;
        });
        return stabilizedThis.map(el => el.el);
    };
    renderShowHideFilter(headCell) {
        if (!headCell.filter) {
            return null;
        }
        return (React.createElement(IconButton, { title: I18n.t('ra_Show/hide filter input'), size: "small", onClick: () => {
                const filterOn = [...this.state.filterOn];
                const pos = this.state.filterOn.indexOf(headCell.attr);
                if (pos === -1) {
                    filterOn.push(headCell.attr);
                }
                else {
                    filterOn.splice(pos, 1);
                }
                this.setState({ filterOn }, () => {
                    if (pos && ConfigTable.getFilterValue(this.filterRefs[headCell.attr])) {
                        ConfigTable.setFilterValue(this.filterRefs[headCell.attr], '');
                        this.applyFilter();
                    }
                });
            } }, this.state.filterOn.includes(headCell.attr) ? React.createElement(IconFilterOff, null) : React.createElement(IconFilterOn, null)));
    }
    renderImportExportButtons(schema) {
        return (React.createElement(React.Fragment, null,
            !schema.noDelete && schema.import ? (React.createElement(Tooltip, { title: I18n.t('ra_Import data from %s file', 'CSV'), slotProps: { popper: { sx: styles.tooltip } } },
                React.createElement(IconButton, { size: "small", onClick: () => this.setState({ showImportDialog: true }) },
                    React.createElement(ImportIcon, null)))) : null,
            schema.export ? (React.createElement(Tooltip, { title: I18n.t('ra_Export data to %s file', 'CSV'), slotProps: { popper: { sx: styles.tooltip } } },
                React.createElement(IconButton, { size: "small", onClick: () => this.onExport() },
                    React.createElement(ExportIcon, null)))) : null,
            schema.noDelete ? null : (React.createElement(IconButton, { disabled: true, size: "small" },
                React.createElement(DeleteIcon, null)))));
    }
    renderAddButton(doAnyFilterSet) {
        return (React.createElement(Tooltip, { title: doAnyFilterSet ? I18n.t('ra_Cannot add items with set filter') : I18n.t('ra_Add row'), slotProps: { popper: { sx: styles.tooltip } } },
            React.createElement("span", null,
                React.createElement(IconButton, { size: "small", color: "primary", disabled: !!doAnyFilterSet && !this.props.schema.allowAddByFilter, onClick: this.onAdd },
                    React.createElement(AddIcon, null)))));
    }
    enhancedTableHead(buttonsWidth, doAnyFilterSet) {
        const { schema } = this.props;
        const { order, orderBy } = this.state;
        return (React.createElement(TableHead, null,
            React.createElement(TableRow, null,
                schema.items?.map((headCell, i) => this.renderOneFilter({
                    schema,
                    style: { width: headCell.width },
                    showAddButton: !i && !schema.noDelete,
                    headCell,
                    order,
                    orderBy,
                    index: i,
                    doAnyFilterSet,
                })),
                !schema.noDelete && (React.createElement(TableCell, { style: {
                        paddingLeft: 20,
                        paddingRight: 20,
                        width: buttonsWidth,
                        textAlign: 'right',
                    }, padding: "checkbox" }, this.renderImportExportButtons(schema))))));
    }
    onDelete = (index) => () => {
        const newValue = JSON.parse(JSON.stringify(this.state.value));
        newValue.splice(index, 1);
        // Clear errors for deleted row and shift remaining error indices
        const newTableErrors = { ...this.state.tableErrors };
        // Clear errors for the deleted row
        if (newTableErrors[index]) {
            // Clear all errors for this row from parent
            Object.keys(newTableErrors[index]).forEach(attr => {
                this.props.onError(attr, undefined);
            });
            delete newTableErrors[index];
        }
        // Shift error indices for rows after the deleted one
        const shiftedErrors = {};
        Object.keys(newTableErrors).forEach(rowIndexStr => {
            const rowIndex = parseInt(rowIndexStr, 10);
            if (rowIndex > index) {
                // Move errors from rowIndex to rowIndex - 1
                shiftedErrors[rowIndex - 1] = newTableErrors[rowIndex];
            }
            else {
                // Keep errors at same index for rows before deleted row
                shiftedErrors[rowIndex] = newTableErrors[rowIndex];
            }
        });
        this.setState({
            value: newValue,
            iteration: this.state.iteration + 10_000,
            tableErrors: shiftedErrors,
        }, () => this.applyFilter(false, null, () => this.onChangeWrapper(newValue)));
    };
    onExport() {
        const { schema } = this.props;
        const { value } = this.state;
        const cols = schema.items.map((it) => it.attr);
        const lines = [cols.join(';')];
        value.forEach(row => {
            const line = [];
            schema.items.forEach((it) => {
                if (row[it.attr] && typeof row[it.attr] === 'string' && row[it.attr].includes(';')) {
                    line.push(`"${row[it.attr]}"`);
                }
                else {
                    line.push(row[it.attr] === undefined || row[it.attr] === null ? '' : row[it.attr]);
                }
            });
            lines.push(line.join(';'));
        });
        const el = document.createElement('a');
        el.setAttribute('href', `data:text/csv;charset=utf-8,${encodeURIComponent(lines.join('\n'))}`);
        const now = new Date();
        el.setAttribute('download', `${now.getFullYear()}_${(now.getMonth() + 1).toString().padStart(2, '0')}_${now.getDate().toString().padStart(2, '0')}_${this.props.oContext.adapterName}.${this.props.oContext.instance}_${this.props.attr}.csv`);
        el.style.display = 'none';
        document.body.appendChild(el);
        el.click();
        document.body.removeChild(el);
    }
    onImport(text) {
        const lines = text.split('\n').map((line) => line.replace('\r', '').trim());
        // the first line is header
        const { schema } = this.props;
        const header = lines
            .shift()
            .split(';')
            .filter(it => it && schema.items.find((it2) => it2.attr === it));
        const values = [];
        lines.forEach((line) => {
            const parts = line.split(';');
            const obj = {};
            for (let p = 0; p < parts.length; p++) {
                let value = parts[p];
                if (value.startsWith('"')) {
                    value = value.substring(1);
                    while (p < parts.length && !value.endsWith('"')) {
                        value += `;${parts[++p]}`;
                    }
                    value = value.substring(0, value.length - 1);
                }
                let val = value;
                if (value === 'true') {
                    val = true;
                }
                else if (value === 'false') {
                    val = false;
                }
                else if (window.isFinite(value)) {
                    const attr = this.props.schema.items.find((it) => it.attr === header[p]);
                    if (attr && attr.type === 'number') {
                        // if a type of attribute is a "number"
                        val = parseFloat(value);
                    }
                    else {
                        val = value;
                    }
                }
                else {
                    val = value;
                }
                obj[header[p]] = val;
            }
            values.push(obj);
        });
        if (values.length) {
            if (this.state.value?.length) {
                this.setState({ showTypeOfImportDialog: values, showImportDialog: false });
            }
            else {
                this.setState({ value: values, showImportDialog: false });
            }
        }
        else {
            window.alert('ra_No data found in file');
        }
    }
    onClone = (index) => () => {
        const newValue = JSON.parse(JSON.stringify(this.state.value));
        const cloned = JSON.parse(JSON.stringify(newValue[index]));
        if (typeof this.props.schema.clone === 'string' && typeof cloned[this.props.schema.clone] === 'string') {
            let i = 1;
            let text = cloned[this.props.schema.clone];
            const pattern = text.match(/(\d+)$/);
            if (pattern) {
                text = text.replace(pattern[0], '');
                i = parseInt(pattern[0], 10) + 1;
            }
            else {
                text += '_';
            }
            while (newValue.find((it) => it[this.props.schema.clone] === text + i.toString())) {
                i++;
            }
            cloned[this.props.schema.clone] = `${cloned[this.props.schema.clone]}_${i}`;
        }
        newValue.splice(index, 0, cloned);
        this.setState({ value: newValue, iteration: this.state.iteration + 10000 }, () => this.applyFilter(false, null, () => this.onChangeWrapper(newValue)));
    };
    onChangeWrapper = (newValue, updateVisible) => {
        if (this.typingTimer) {
            clearTimeout(this.typingTimer);
        }
        this.typingTimer = setTimeout((value, _updateVisible) => {
            this.typingTimer = null;
            let mayBePromise;
            if (this.props.schema.encryptedAttributes) {
                const _value = JSON.parse(JSON.stringify(value));
                _value.forEach((el) => {
                    this.props.schema.encryptedAttributes.forEach((attr) => {
                        if (el[attr]) {
                            el[attr] = encrypt(this.secret, el[attr]);
                        }
                    });
                });
                if (this.props.schema.objKeyName) {
                    const objValue = arrayToObject(_value, this.props.schema.objKeyName, this.props.schema.objValueName);
                    mayBePromise = this.onChange(this.props.attr, objValue);
                }
                else {
                    mayBePromise = this.onChange(this.props.attr, _value);
                }
            }
            else if (this.props.schema.objKeyName) {
                const objValue = arrayToObject(JSON.parse(JSON.stringify(value)), this.props.schema.objKeyName, this.props.schema.objValueName);
                mayBePromise = this.onChange(this.props.attr, objValue);
            }
            else {
                mayBePromise = this.onChange(this.props.attr, value);
            }
            if (mayBePromise instanceof Promise) {
                mayBePromise.catch(e => console.error(`Cannot save: ${e}`));
            }
            if (_updateVisible) {
                this.applyFilter(false, value);
                this.handleRequestSort(this.state.orderBy, true);
            }
        }, 300, newValue, updateVisible);
    };
    onAdd = () => {
        const { schema } = this.props;
        const newValue = JSON.parse(JSON.stringify(this.state.value));
        const newItem = schema.items?.reduce((accumulator, currentValue) => {
            let defaultValue;
            if (currentValue.defaultFunc) {
                if (this.props.custom) {
                    defaultValue = currentValue.defaultFunc
                        ? this.executeCustom(currentValue.defaultFunc, this.props.data, this.props.customObj, this.props.oContext.instanceObj, newValue.length, this.props.data)
                        : this.props.schema.default;
                }
                else {
                    defaultValue = currentValue.defaultFunc
                        ? this.execute(currentValue.defaultFunc, this.props.schema.default, this.props.data, newValue.length, this.props.data)
                        : this.props.schema.default;
                }
            }
            else {
                defaultValue = currentValue.default === undefined ? null : currentValue.default;
            }
            accumulator[currentValue.attr] = defaultValue;
            return accumulator;
        }, {});
        newValue.push(newItem);
        this.setState({ value: newValue }, () => this.applyFilter(false, null, () => this.onChangeWrapper(newValue)));
    };
    isAnyFilterSet() {
        return !!Object.keys(this.filterRefs).find(attr => ConfigTable.getFilterValue(this.filterRefs[attr]));
    }
    applyFilter = (clear, value, cb) => {
        value = value || this.state.value;
        let visibleValue = value.map((_, i) => i);
        Object.keys(this.filterRefs).forEach(attr => {
            let valueInputRef = ConfigTable.getFilterValue(this.filterRefs[attr]);
            if (!clear && valueInputRef) {
                valueInputRef = valueInputRef.toLowerCase();
                visibleValue = visibleValue.filter(idx => value[idx] && value[idx][attr] && value[idx][attr].toLowerCase().includes(valueInputRef));
            }
            else if (this.filterRefs[attr].current) {
                ConfigTable.setFilterValue(this.filterRefs[attr], '');
            }
        });
        if (visibleValue.length === value.length) {
            visibleValue = null;
        }
        if (visibleValue === null && this.state.visibleValue === null) {
            if (cb) {
                cb();
            }
            return;
        }
        if (JSON.stringify(visibleValue) !== JSON.stringify(this.state.visibleValue)) {
            this.setState({ visibleValue }, () => cb && cb());
        }
        else if (cb) {
            cb();
        }
    };
    onMoveUp(idx) {
        const newValue = JSON.parse(JSON.stringify(this.state.value));
        const item = newValue[idx];
        newValue.splice(idx, 1);
        newValue.splice(idx - 1, 0, item);
        this.setState({ value: newValue, iteration: this.state.iteration + 10000 }, () => this.applyFilter(false, null, () => this.onChangeWrapper(newValue)));
    }
    onMoveDown(idx) {
        const newValue = JSON.parse(JSON.stringify(this.state.value));
        const item = newValue[idx];
        newValue.splice(idx, 1);
        newValue.splice(idx + 1, 0, item);
        this.setState({ value: newValue, iteration: this.state.iteration + 10000 }, () => this.applyFilter(false, null, () => this.onChangeWrapper(newValue)));
    }
    onDrop(acceptedFiles) {
        const file = acceptedFiles[0];
        const reader = new FileReader();
        reader.onabort = () => console.log('file reading was aborted');
        reader.onerror = () => console.log('file reading has failed');
        reader.onload = () => {
            if (file.size > MAX_SIZE) {
                window.alert(I18n.t('ra_File is too big. Max %sk allowed. Try use SVG.', Math.round(MAX_SIZE / 1024)));
                return;
            }
            const text = new Uint8Array(reader.result).reduce((data, byte) => data + String.fromCharCode(byte), '');
            this.onImport(text);
        };
        reader.readAsArrayBuffer(file);
    }
    showTypeOfImportDialog() {
        if (!this.state.showTypeOfImportDialog) {
            return null;
        }
        return (React.createElement(Dialog, { open: !0, onClose: () => this.setState({ showTypeOfImportDialog: false }), maxWidth: "md" },
            React.createElement(DialogTitle, null, I18n.t('ra_Append or replace?')),
            React.createElement(DialogContent, null, I18n.t('ra_Append %s entries or replace existing?', this.state.showTypeOfImportDialog.length)),
            React.createElement(DialogActions, null,
                React.createElement(Button, { variant: "contained", color: "primary", autoFocus: true, onClick: () => {
                        const value = JSON.parse(JSON.stringify(this.state.value));
                        this.state.showTypeOfImportDialog.forEach((obj) => value.push(obj));
                        this.setState({
                            value,
                            iteration: this.state.iteration + 10000,
                            showTypeOfImportDialog: false,
                        }, () => this.applyFilter(false, null, () => this.onChangeWrapper(value)));
                    } }, I18n.t('ra_Append')),
                React.createElement(Button, { variant: "contained", color: "secondary", autoFocus: true, onClick: () => {
                        const value = this.state.showTypeOfImportDialog;
                        this.setState({
                            value,
                            iteration: this.state.iteration + 10000,
                            showTypeOfImportDialog: false,
                        }, () => this.applyFilter(false, null, () => this.onChangeWrapper(value)));
                    } }, I18n.t('ra_Replace')))));
    }
    showImportDialog() {
        if (!this.state.showImportDialog) {
            return null;
        }
        return (React.createElement(Dialog, { open: !0, onClose: () => this.setState({ showImportDialog: false }), sx: {
                '& .MuiDialog-paper': {
                    minHeight: 500,
                },
            }, maxWidth: "md", fullWidth: true },
            React.createElement(DialogTitle, null, I18n.t('ra_Import from %s', 'CSV')),
            React.createElement(DialogContent, null,
                React.createElement(Dropzone, { multiple: false, accept: { 'text/csv': ['.csv'] }, maxSize: MAX_SIZE, onDragEnter: () => this.setState({ uploadFile: 'dragging' }), onDragLeave: () => this.setState({ uploadFile: true }), onDrop: (acceptedFiles, errors) => {
                        this.setState({ uploadFile: false });
                        if (!acceptedFiles.length) {
                            window.alert((errors &&
                                errors[0] &&
                                errors[0].errors &&
                                errors[0].errors[0] &&
                                errors[0].errors[0].message) ||
                                I18n.t('ra_Cannot upload'));
                        }
                        else {
                            this.onDrop(acceptedFiles);
                        }
                    } }, ({ getRootProps, getInputProps }) => (React.createElement("div", { style: {
                        ...styles.uploadDiv,
                        ...(this.state.uploadFile === 'dragging' ? styles.uploadDivDragging : undefined),
                        ...styles.dropZone,
                        ...(!this.state.icon ? styles.dropZoneEmpty : undefined),
                    }, ...getRootProps() },
                    React.createElement("input", { ...getInputProps() }),
                    React.createElement("div", { style: styles.uploadCenterDiv },
                        React.createElement("div", { style: styles.uploadCenterTextAndIcon },
                            React.createElement(ImportIcon, { style: styles.uploadCenterIcon }),
                            React.createElement("div", { style: styles.uploadCenterText }, this.state.uploadFile === 'dragging'
                                ? I18n.t('ra_Drop file here')
                                : I18n.t('ra_Place your files here or click here to open the browse dialog')))))))),
            React.createElement(DialogActions, null,
                React.createElement(Button, { variant: "contained", onClick: () => this.setState({ showImportDialog: false }), color: "primary", startIcon: React.createElement(IconClose, null) }, I18n.t('Cancel')))));
    }
    renderOneFilter(props) {
        return (React.createElement(TableCell, { key: `${props.headCell.attr}_${props.index}`, style: props.style, align: "left", sortDirection: props.orderBy === props.headCell.attr ? props.order : false },
            React.createElement("div", { style: {
                    ...styles.flex,
                    ...(props.schema.showFirstAddOnTop ? { flexDirection: 'column' } : undefined),
                } },
                props.showAddButton ? this.renderAddButton(props.doAnyFilterSet) : null,
                props.headCell.sort && (React.createElement(TableSortLabel, { active: true, style: props.orderBy !== props.headCell.attr ? styles.silver : undefined, direction: props.orderBy === props.headCell.attr ? props.order : 'asc', onClick: () => this.handleRequestSort(props.headCell.attr) })),
                props.headCell.filter && this.state.filterOn.includes(props.headCell.attr) ? (React.createElement(TextField, { variant: "standard", ref: this.filterRefs[props.headCell.attr], onChange: () => this.applyFilter(), title: I18n.t('ra_You can filter entries by entering here some text'), slotProps: {
                        input: {
                            endAdornment: ConfigTable.getFilterValue(this.filterRefs[props.headCell.attr]) && (React.createElement(InputAdornment, { position: "end" },
                                React.createElement(IconButton, { size: "small", tabIndex: -1, onClick: () => {
                                        ConfigTable.setFilterValue(this.filterRefs[props.headCell.attr], '');
                                        this.applyFilter();
                                    } },
                                    React.createElement(CloseIcon, null)))),
                        },
                    }, fullWidth: true, placeholder: this.getText(props.headCell.title) })) : (React.createElement("span", { style: styles.headerText }, this.getText(props.headCell.title))),
                this.renderShowHideFilter(props.headCell))));
    }
    enhancedFilterCard() {
        const { schema } = this.props;
        const { order, orderBy } = this.state;
        let tdStyle;
        if (this.props.schema.compact) {
            tdStyle = { paddingTop: 1, paddingBottom: 1 };
        }
        const importExportVisible = (!schema.noDelete && schema.import) || schema.export;
        if (importExportVisible || schema.items.find(item => item.sort || item.filter)) {
            return (React.createElement(Grid2, { size: {
                    xs: schema.xs || 12, // if xs is not defined, take the full width
                    sm: schema.sm || undefined,
                    md: schema.md || undefined,
                    lg: schema.lg || undefined,
                    xl: schema.xl || undefined,
                } },
                React.createElement(Card, null,
                    React.createElement(Paper, { style: styles.paper },
                        React.createElement(Accordion, { style: styles.paper },
                            React.createElement(AccordionSummary, { expandIcon: React.createElement(ExpandMoreIcon, null) },
                                React.createElement(Typography, null, I18n.t('ra_Filter and Data Actions'))),
                            React.createElement(AccordionDetails, null,
                                React.createElement(Table, null,
                                    React.createElement(TableBody, null,
                                        schema.items?.find(item => item.sort || item.filter) &&
                                            schema.items.map((headCell, i) => (React.createElement(TableRow, { key: `${headCell.attr}_${i}` }, this.renderOneFilter({
                                                schema,
                                                style: tdStyle,
                                                showAddButton: false,
                                                headCell,
                                                order,
                                                orderBy,
                                                index: i,
                                                doAnyFilterSet: false,
                                            })))),
                                        importExportVisible ? (React.createElement(TableRow, null,
                                            React.createElement(TableCell, { align: "left", style: tdStyle },
                                                React.createElement("span", { style: styles.headerText }, I18n.t('ra_Actions'))),
                                            React.createElement(TableCell, { style: tdStyle }, this.renderImportExportButtons(schema)))) : null))))))));
        }
        return null;
    }
    enhancedBottomCard() {
        const { schema } = this.props;
        let tdStyle;
        if (this.props.schema.compact) {
            tdStyle = { paddingTop: 1, paddingBottom: 1 };
        }
        const doAnyFilterSet = this.isAnyFilterSet();
        return (React.createElement(Grid2, { size: {
                xs: schema.xs || 12, // if xs is not defined, take the full width
                sm: schema.sm || undefined,
                md: schema.md || undefined,
                lg: schema.lg || undefined,
                xl: schema.xl || undefined,
            } },
            React.createElement(Card, null,
                React.createElement(Paper, { style: styles.paper },
                    React.createElement(Table, null,
                        React.createElement(TableBody, null,
                            React.createElement(TableRow, null,
                                React.createElement(TableCell, { colSpan: schema.items.length + 1, style: tdStyle }, this.renderAddButton(doAnyFilterSet)))))))));
    }
    renderCards() {
        const { schema } = this.props;
        let { visibleValue } = this.state;
        let tdStyle;
        if (this.props.schema.compact) {
            tdStyle = { paddingTop: 1, paddingBottom: 1 };
        }
        visibleValue ||= this.state.value.map((_, i) => i);
        const doAnyFilterSet = this.isAnyFilterSet();
        return (React.createElement(Grid2, { container: true },
            this.showImportDialog(),
            this.showTypeOfImportDialog(),
            this.enhancedFilterCard(),
            visibleValue.map((idx, i) => (React.createElement(Grid2, { key: `${idx}_${i}`, size: {
                    xs: schema.xs || 12, // if xs is not defined, take the full width
                    sm: schema.sm || undefined,
                    md: schema.md || undefined,
                    lg: schema.lg || undefined,
                    xl: schema.xl || undefined,
                } },
                React.createElement(Card, null,
                    React.createElement(Paper, { style: styles.paper },
                        React.createElement(Table, null,
                            React.createElement(TableBody, null,
                                schema.items?.map((headCell) => (React.createElement(TableRow, { key: `${headCell.attr}_${idx}` },
                                    React.createElement(TableCell, { align: "left", style: tdStyle },
                                        React.createElement("span", { style: styles.headerText }, this.getText(headCell.title))),
                                    React.createElement(TableCell, { align: "left", style: tdStyle }, this.itemTable(headCell.attr, this.state.value[idx], idx))))),
                                !schema.noDelete && (React.createElement(TableRow, null,
                                    React.createElement(TableCell, { align: "left", style: tdStyle },
                                        React.createElement("span", { style: styles.headerText }, this.getText('Actions'))),
                                    React.createElement(TableCell, { align: "left", style: tdStyle },
                                        !doAnyFilterSet && !this.state.orderBy ? (React.createElement(Tooltip, { title: I18n.t('ra_Move up'), slotProps: { popper: { sx: styles.tooltip } } },
                                            React.createElement("span", null,
                                                React.createElement(IconButton, { size: "small", onClick: () => this.onMoveUp(idx), disabled: i === 0 },
                                                    React.createElement(UpIcon, null))))) : null,
                                        !doAnyFilterSet && !this.state.orderBy ? (React.createElement(Tooltip, { title: I18n.t('ra_Move down'), slotProps: { popper: { sx: styles.tooltip } } },
                                            React.createElement("span", null,
                                                React.createElement(IconButton, { size: "small", onClick: () => this.onMoveDown(idx), disabled: i === visibleValue.length - 1 },
                                                    React.createElement(DownIcon, null))))) : null,
                                        React.createElement(Tooltip, { title: I18n.t('ra_Delete current row'), slotProps: { popper: { sx: styles.tooltip } } },
                                            React.createElement(IconButton, { size: "small", onClick: this.onDelete(idx) },
                                                React.createElement(DeleteIcon, null))),
                                        this.props.schema.clone ? (React.createElement(Tooltip, { title: I18n.t('ra_Clone current row'), slotProps: { popper: { sx: styles.tooltip } } },
                                            React.createElement(IconButton, { size: "small", onClick: this.onClone(idx) },
                                                React.createElement(CopyContentIcon, null)))) : null)))))))))),
            this.enhancedBottomCard()));
    }
    renderTable() {
        const { schema } = this.props;
        let { visibleValue } = this.state;
        visibleValue = visibleValue || this.state.value.map((_, i) => i);
        const doAnyFilterSet = this.isAnyFilterSet();
        let tdStyle;
        if (this.props.schema.compact) {
            tdStyle = { paddingTop: 1, paddingBottom: 1 };
        }
        return (React.createElement(Paper, { style: styles.paper },
            this.showImportDialog(),
            this.showTypeOfImportDialog(),
            schema.label ? (React.createElement("div", { style: styles.label },
                React.createElement(Toolbar, { variant: "dense", style: styles.rootTool },
                    React.createElement(Typography, { style: styles.title, variant: "h6", id: "tableTitle", component: "div" }, this.getText(schema.label))))) : null,
            React.createElement(TableContainer, null,
                React.createElement(Table, { style: styles.table, size: "small" },
                    this.enhancedTableHead(!doAnyFilterSet && !this.state.orderBy ? 120 : 64, doAnyFilterSet),
                    React.createElement(TableBody, null,
                        visibleValue.map((idx, i) => (React.createElement(TableRow, { hover: true, key: `${idx}_${i}` },
                            schema.items?.map((headCell) => (React.createElement(TableCell, { key: `${headCell.attr}_${idx}`, align: "left", style: tdStyle }, this.itemTable(headCell.attr, this.state.value[idx], idx)))),
                            !schema.noDelete && (React.createElement(TableCell, { align: "left", style: { ...tdStyle, ...styles.buttonCell } },
                                !doAnyFilterSet && !this.state.orderBy ? (i ? (React.createElement(Tooltip, { title: I18n.t('ra_Move up'), slotProps: { popper: { sx: styles.tooltip } } },
                                    React.createElement(IconButton, { size: "small", onClick: () => this.onMoveUp(idx) },
                                        React.createElement(UpIcon, null)))) : (React.createElement("div", { style: styles.buttonEmpty }))) : null,
                                !doAnyFilterSet && !this.state.orderBy ? (i < visibleValue.length - 1 ? (React.createElement(Tooltip, { title: I18n.t('ra_Move down'), slotProps: { popper: { sx: styles.tooltip } } },
                                    React.createElement(IconButton, { size: "small", onClick: () => this.onMoveDown(idx) },
                                        React.createElement(DownIcon, null)))) : (React.createElement("div", { style: styles.buttonEmpty }))) : null,
                                React.createElement(Tooltip, { title: I18n.t('ra_Delete current row'), slotProps: { popper: { sx: styles.tooltip } } },
                                    React.createElement(IconButton, { size: "small", onClick: this.onDelete(idx) },
                                        React.createElement(DeleteIcon, null))),
                                this.props.schema.clone ? (React.createElement(Tooltip, { title: I18n.t('ra_Clone current row'), slotProps: { popper: { sx: styles.tooltip } } },
                                    React.createElement(IconButton, { size: "small", onClick: this.onClone(idx) },
                                        React.createElement(CopyContentIcon, null)))) : null))))),
                        !schema.noDelete && visibleValue.length >= (schema.showSecondAddAt || 5) ? (React.createElement(TableRow, null,
                            React.createElement(TableCell, { colSpan: schema.items.length + 1, style: { ...tdStyle } }, this.renderAddButton(doAnyFilterSet)))) : null)),
                !visibleValue.length && this.state.value.length ? (React.createElement("div", { style: styles.filteredOut },
                    React.createElement(Typography, { style: styles.title, variant: "h6", id: "tableTitle", component: "div" },
                        I18n.t('ra_All items are filtered out'),
                        React.createElement(IconButton, { size: "small", onClick: () => this.applyFilter(true) },
                            React.createElement(CloseIcon, null))))) : null),
            schema.help ? (React.createElement(FormHelperText, { style: { paddingLeft: 16 } }, this.renderHelp(this.props.schema.help, this.props.schema.helpLink, this.props.schema.noTranslation))) : null,
            this.state.errorMessage ? (React.createElement("div", { style: { display: 'flex', padding: '5px' } },
                React.createElement(ErrorIcon, { color: "error" }),
                React.createElement("span", { style: { color: 'red', alignSelf: 'center' } }, this.state.errorMessage))) : null));
    }
    componentDidUpdate() {
        if (this.refDiv.current?.clientWidth && this.refDiv.current.clientWidth !== this.state.width) {
            if (this.resizeTimeout) {
                clearTimeout(this.resizeTimeout);
            }
            this.resizeTimeout = setTimeout(() => {
                this.resizeTimeout = null;
                this.setState({ width: this.refDiv.current?.clientWidth });
            }, 50);
        }
    }
    getCurrentBreakpoint() {
        if (!this.state.width) {
            return 'md';
        }
        if (this.state.width < 600) {
            return 'xs';
        }
        if (this.state.width < 900) {
            return 'sm';
        }
        if (this.state.width < 1200) {
            return 'md';
        }
        if (this.state.width < 1536) {
            return 'lg';
        }
        return 'xl';
    }
    renderItem( /* error, disabled, defaultValue */) {
        const { schema } = this.props;
        if (!this.state.value || !Array.isArray(this.state.value)) {
            return null;
        }
        const currentBreakpoint = this.getCurrentBreakpoint();
        let content;
        if (currentBreakpoint && (schema.useCardFor || ['xs']).includes(currentBreakpoint)) {
            content = this.renderCards();
        }
        else {
            content = this.renderTable();
        }
        return (React.createElement("div", { ref: this.refDiv, style: { width: '100%' } }, content));
    }
}
export default ConfigTable;
